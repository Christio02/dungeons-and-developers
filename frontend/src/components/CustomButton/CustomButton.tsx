import { Link } from 'react-router-dom';

interface ButtonProps {
  text: string;
  onClick?: () => void;
  className?: string;
  textSize?: string;
  linkTo?: string;
  noUnderline?: boolean;
  isActive?: boolean;
  children?: React.ReactNode;
}

export default function CustomButton({
                                       text,
                                       onClick,
                                       className = 'sub-header',
                                       linkTo,
                                       noUnderline = false,
                                       isActive = false,
                                       children,
                                     }: ButtonProps) {
  const underlineClass = noUnderline
    ? ''
    : `absolute bottom-0 w-0 h-0.5 bg-white transition-all duration-300 group-hover:w-full group-hover:left-0 ${
      isActive ? 'w-full left-0' : 'left-1/2' // Underline for active page on desktop
    }`;

  const activeClass = isActive
    ? 'bold xl:font-[MedievalSharp]' // Bold text for active page on mobile
    : '';

  const buttonContent = (
    <>
      {text}
      <span className={underlineClass}></span>
      {children}
    </>
  );

  // If linkTo is provided, use Link component for navigation
  if (linkTo) {
    return (
      <Link to={linkTo} className={`w-fit relative group flex ${activeClass} ${className}`}>
        {buttonContent}
      </Link>
    );
  }

  // Otherwise, render a regular button
  return (
    <button className={`relative group flex ${activeClass} ${className}`} onClick={onClick}>
      {buttonContent}
    </button>
  );
}