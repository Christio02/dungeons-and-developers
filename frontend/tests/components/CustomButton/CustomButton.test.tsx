import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CustomButton from '../../../src/components/CustomButton/CustomButton.tsx';
import { MemoryRouter } from 'react-router-dom';
import { describe, it, expect, vi } from 'vitest';

describe('CustomButton', () => {
  it('renders correctly as a button and matches snapshot', () => {
    const { container } = render(<CustomButton text="Click me" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly as a Link and matches snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <CustomButton text="Go Home" linkTo="/home" />
      </MemoryRouter>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('handles click events', async () => {
    const handleClick = vi.fn();
    render(<CustomButton text="Click me" onClick={handleClick} />);
    const button = screen.getByText(/Click me/i);

    await userEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('renders with custom className and children', () => {
    render(
      <CustomButton text="Click me" className="custom-class">
        <span>Child Element</span>
      </CustomButton>
    );

    const button = screen.getByText(/Click me/i);
    expect(button).toHaveClass('custom-class');
    expect(screen.getByText(/Child Element/i)).toBeInTheDocument();
  });

  it('renders an active button with underline', () => {
    render(<CustomButton text="Active" isActive />);
    const button = screen.getByText(/Active/i);
    expect(button).toHaveClass('bold');
  });

  it('handles focus and blur events', async () => {
    const handleFocus = vi.fn();
    const handleBlur = vi.fn();

    render(<CustomButton text="Focus Test" onFocus={handleFocus} onBlur={handleBlur} />);

    const button = screen.getByText(/Focus Test/i);
    await userEvent.tab(); // Simulate focus
    expect(handleFocus).toHaveBeenCalledTimes(1);

    await userEvent.tab(); // Simulate blur
    expect(handleBlur).toHaveBeenCalledTimes(1);
  });

  it('navigates correctly when used as a Link', () => {
    render(
      <MemoryRouter>
        <CustomButton text="Go to Page" linkTo="/page" />
      </MemoryRouter>
    );

    const link = screen.getByText(/Go to Page/i);
    expect(link.tagName).toBe('A');
    expect(link).toHaveAttribute('href', '/page');
  });
});
