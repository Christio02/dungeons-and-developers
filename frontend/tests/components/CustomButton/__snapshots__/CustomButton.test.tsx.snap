// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomButton > renders correctly as a Link and matches snapshot 1`] = `
<a
  class="w-fit relative group flex sub-header  sub-header"
  href="/home"
>
  Go Home
  <span
    class="absolute bottom-0 w-0 h-0.5 bg-white transition-all duration-300 group-hover:w-full group-hover:left-0 left-1/2"
  />
</a>
`;

exports[`CustomButton > renders correctly as a button and matches snapshot 1`] = `
<button
  class="relative group flex sub-header  sub-header"
>
  Click me
  <span
    class="absolute bottom-0 w-0 h-0.5 bg-white transition-all duration-300 group-hover:w-full group-hover:left-0 left-1/2"
  />
</button>
`;
